lib/puppet/transaction/report.rb

```diff
@@ -377,9 +377,15 @@ def raw_summary
   #
   def exit_status
     status = 0
-    status |= 2 if @metrics["changes"]["total"] > 0
-    status |= 4 if @metrics["resources"]["failed"] > 0
-    status |= 4 if @metrics["resources"]["failed_to_restart"] > 0
+    if @metrics["changes"] && @metrics["changes"]["total"] &&
+        @metrics["resources"] && @metrics["resources"]["failed"] &&
+        @metrics["resources"]["failed_to_restart"]
+      status |= 2 if @metrics["changes"]["total"] > 0
+      status |= 4 if @metrics["resources"]["failed"] > 0
+      status |= 4 if @metrics["resources"]["failed_to_restart"] > 0
+    else
+      status = -1
```

Comment:
I'd stay away from -1 as it has some loosely defined semantics. Unix process exit codes are 8 bit unsigned numbers so this'll be converted to 255, and an exit code of 255 indicates that the exit code is out of range. Strictly speaking an out of range exit code is returned modulo 256, so 255 isn't specifically reserved, but per http://tldp.org/LDP/abs/html/exitcodes.html people largely assume that 255 is special.

That said, exiting with -1/255 wouldn't be terrible or burn the world down or anything drastic like that, so we could stick with this approach. However if we have the ability to change this to another exit value that's a bit less strange we might want to use this opportunity accordingly.